src/infra/ingest-datafusion/src/writer.rs

```diff
@@ -851,16 +854,122 @@ impl DataWriterDataFusionBuilder {
             .resolve_ref(&self.block_ref)
             .await
             .int_err()?;
+        let mut seed_visitor = SearchSeedVisitor::<ScanMetadataError>::default();
+        let mut set_vocab_visitor = SearchSetVocabVisitor::<ScanMetadataError>::default();
+        let mut set_data_schema_visitor =
+            SearchSetDataSchemaVisitor::<ScanMetadataError>::default();
+        let mut prev_source_state_visitor = GenericCallbackVisitor::new(
+            None::<SourceState>,
+            Decision::NextOfType(Flag::ADD_DATA),
+            |state, _, block| {
+                let MetadataEvent::AddData(e) = &block.event else {
+                    unreachable!()
+                };
+
+                if let Some(ss) = &e.new_source_state
+                    && let Some(sn) = source_name
+                    && sn != ss.source_name.as_str()
+                {
+                    unimplemented!(
+                        "Differentiating between the state of multiple sources is not yet \
+                         supported"
+                    );
+                }
+
+                *state = e.new_source_state.clone();
+
+                Ok(Decision::Stop)
+            },
+        );
+
+        struct AddDataCollectionVisitorState {
+            data_slices: Vec<Multihash>,
+            prev_checkpoint: Option<Multihash>,
+            prev_watermark: Option<DateTime<Utc>>,
+            prev_offset: Option<u64>,
+        }
 
-        let (decision, mut visitor) =
-            DataWriterDataFusionMetaDataStateVisitor::new(head.clone(), source_name);
+        let mut add_data_collection_visitor = GenericCallbackVisitor::new(
+            AddDataCollectionVisitorState {
+                data_slices: Vec::new(),
+                prev_checkpoint: None,
+                prev_watermark: None,
+                prev_offset: None,
```

Comment:
Might be simpler to separate these into a visitor that gets last `AddDataBlock` - this way you won't need all the `is_none()` checks below.